### PR TITLE
Fix aggregate type coercion bug

### PR DESCRIPTION
**Pull request by alamb**
~DRAFT -- I am just pushing it up in case anyone else finds the debugging via logging helpful~

# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/3704

 # Rationale for this change
There is a planning regression

# What changes are included in this PR?
Fix bug (will comment inline)

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/optimizer/src/optimizer.rs [hunk 1]
```diff
@@ -178,16 +178,15 @@ impl Optimizer {
         F: FnMut(&LogicalPlan, &dyn OptimizerRule),
     {
         let mut new_plan = plan.clone();
-        debug!("Input logical plan:\n{}\n", plan.display_indent());
-        trace!("Full input logical plan:\n{:?}", plan);
+        log_plan("Optimizer input", plan);
```
Comment (alamb) [inline datafusion/optimizer/src/optimizer.rs:181]: This is a driveby cleanup to improve logging (specifically, also add `trace!` to log schema)

File: datafusion/core/tests/sql/subqueries.rs [hunk 2]
```diff
@@ -147,8 +147,8 @@ order by s_acctbal desc, n_name, s_name, p_partkey;"#;
           Inner Join: #supplier.s_nationkey = #nation.n_nationkey
             Inner Join: #partsupp.ps_suppkey = #supplier.s_suppkey
               Inner Join: #part.p_partkey = #partsupp.ps_partkey
-                Filter: #part.p_size = Int32(15) AND #part.p_type LIKE Utf8("%BRASS")
-                  TableScan: part projection=[p_partkey, p_mfgr, p_type, p_size], partial_filters=[#part.p_size = Int32(15), #part.p_type LIKE Utf8("%BRASS")]
+                Filter: CAST(#part.p_size AS Int64) = Int64(15) AND #part.p_type LIKE Utf8("%BRASS")
```
Comment (alamb) [inline datafusion/core/tests/sql/subqueries.rs:150]: to be honest, I am not sure why this has changed (aka the filters are no longer simplified). I will look into that in the morning

File: datafusion/optimizer/src/unwrap_cast_in_comparison.rs [hunk 3]
```diff
@@ -97,7 +97,18 @@ fn optimize(plan: &LogicalPlan) -> Result<LogicalPlan> {
     let new_exprs = plan
         .expressions()
         .into_iter()
-        .map(|expr| expr.rewrite(&mut expr_rewriter))
+        .map(|expr| {
+            let original_name = expr.name()?;
```
Comment (alamb) [inline datafusion/optimizer/src/unwrap_cast_in_comparison.rs:101]: I think the root cause issue is that the `UnwrapCastInComparison` can add potentially change the expression but it doesn't add an alias so the output name changes

Bigger picture there are at least three places that we have rediscovered this same problem when rewriting expressions --https://github.com/apache/arrow-datafusion/issues/3555  and https://github.com/apache/arrow-datafusion/blob/master/datafusion/optimizer/src/simplify_expressions.rs#L316  I will try and make a follow on PR to clean them all up. In particular, I think this is something `from_plan` could potentially handle

Comment (andygrove) [reply datafusion/optimizer/src/unwrap_cast_in_comparison.rs:101]: `expr.name()` here hides the casts that are added by this optimization rule, so `expr.name()` is the same as the original name (even though the expression is now different), and the alias does not get added.

Comment (alamb) [reply datafusion/optimizer/src/unwrap_cast_in_comparison.rs:101]: I think this optimizer rule may actually change the name (e.g. from `Int64(0)` to `Int32(0)`) which i think is the root cause of the issue in this bug

File: datafusion/optimizer/tests/integration-test.rs [hunk 4]
```diff
@@ -46,6 +52,17 @@ fn case_when() -> Result<()> {
     Ok(())
 }
 
+#[test]
+fn case_when_aggregate() -> Result<()> {
+    let sql = "SELECT col_utf8, SUM(CASE WHEN col_int32 > 0 THEN 1 ELSE 0 END) AS n FROM test GROUP BY col_utf8";
+    let plan = test_sql(sql)?;
+    let expected = "Projection: #test.col_utf8, #SUM(CASE WHEN test.col_int32 > Int64(0) THEN Int64(1) ELSE Int64(0) END) AS n\
```
Comment (andygrove) [inline datafusion/optimizer/tests/integration-test.rs:59]: Could you merge latest from master - we should not include `#` before column names now

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @alamb

File: datafusion/optimizer/tests/integration-test.rs (hunk 4 above)
Comment (alamb) [reply datafusion/optimizer/tests/integration-test.rs:59]: I have fixed the issue -- it was rewriting sorts. New changes very close

Comment (alamb) [reply datafusion/optimizer/tests/integration-test.rs:59]: Will do

Comment (alamb) [reply datafusion/optimizer/tests/integration-test.rs]: in e3830a7bd

File: datafusion/optimizer/tests/integration-test.rs [hunk 5]
```diff
@@ -29,13 +29,19 @@ use std::any::Any;
...
+}
+
 #[test]
 fn case_when() -> Result<()> {
     let sql = "SELECT CASE WHEN col_int32 > 0 THEN 1 ELSE 0 END FROM test";
     let plan = test_sql(sql)?;
     let expected =
-        "Projection: CASE WHEN test.col_int32 > Int32(0) THEN Int64(1) ELSE Int64(0) END\
-    \n  TableScan: test projection=[col_int32]";
+        "Projection: CASE WHEN test.col_int32 > Int32(0) THEN Int64(1) ELSE Int64(0) END AS CASE WHEN test.col_int32 > Int64(0) THEN Int64(1) ELSE Int64(0) END\
```
Comment (alamb) [inline datafusion/optimizer/tests/integration-test.rs:43]: FYI @andygrove  the alias was added to this as well

File: datafusion/optimizer/src/unwrap_cast_in_comparison.rs [hunk 6]
```diff
@@ -97,12 +97,47 @@ fn optimize(plan: &LogicalPlan) -> Result<LogicalPlan> {
...
+            let original_name = name_for_alias(&expr)?;
+            let expr = expr.rewrite(&mut expr_rewriter)?;
+            add_alias_if_changed(&original_name, expr)
+        })
         .collect::<Result<Vec<_>>>()?;
 
     from_plan(plan, new_exprs.as_slice(), new_inputs.as_slice())
 }
 
+fn name_for_alias(expr: &Expr) -> Result<String> {
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/optimizer/src/unwrap_cast_in_comparison.rs:110]: follow on https://github.com/apache/arrow-datafusion/pull/3727

**Comment by andygrove** [comment]
Thanks again @alamb. This was a huge improvement.

**Comment by alamb** [comment]
I am pretty excited where the datafusion optimizer is headign
